soc/xtensa/intel_adsp/common/include/cavs_hda.h

```diff
@@ -0,0 +1,303 @@
+ */
+#define cavs_hda_dbg(name, base, sid)					\
+		printk("%s:%u %s(%u:0x%p), dgcs: 0x%x, dgbba 0x%x, "	\
+		       "dgbs %u, dgbrp %u, dgbwp %u, dgbsp %u, "	\
+		       "dgmbs %u, dgbllpi 0x%x, dglpibi 0x%x\n",	\
+		       __FILE__, __LINE__, name,			\
+		       sid, DGCS(base, sid),				\
+		       *DGCS(base, sid),				\
+		       *DGBBA(base, sid),				\
+		       *DGBS(base, sid),				\
+		       *DGBRP(base, sid),				\
+		       *DGBWP(base, sid),				\
+		       *DGBSP(base, sid),				\
+		       *DGMBS(base, sid),				\
+		       *DGLLPI(base, sid),				\
+		       *DGLPIBI(base, sid))
+
+
+
+/**
+ * @brief Initialize an HDA stream for use with the firmware
+ *
+ * @param hda Stream set to work with
+ * @param sid Stream ID
+ */
+static inline void cavs_hda_init(uint32_t base, uint32_t sid)
+{
+	*DGCS(base, sid) |= DGCS_FWCB;
+}
+
+/**
+ * @brief Set the buffer, size, and element size for an HDA stream
+ *
+ * Sanity checks that the buffer address and size are valid and that the
+ * stream isn't enabled or busy.
+ *
+ * Prior to enabling an HDA stream to/from the host this is the minimum configuration
+ * that is required. It must be set *after* the host has configured its own buffers.
+ *
+ *
+ * @param hda Stream set to work with
+ * @param sid Stream ID
+ * @param buf Buffer address to use for the shared FIFO. Must be in L2 and 128 byte aligned.
+ * @param buf_size Buffer size in bytes Must be 128 byte aligned
+ *
+ * @retval -EBUSY if the HDA stream is already enabled
+ * @retval -EINVAL if the buf is not in L2, buf isn't aligned on 128 byte boundaries
+ * @retval 0 on Success
+ */
+static inline int cavs_hda_set_buffer(uint32_t base, uint32_t sid,
+				     uint8_t *buf, uint32_t buf_size)
+{
+	void *cached_buf = arch_xtensa_cached_ptr(buf);
+	uint32_t cached_addr = (uint32_t)cached_buf;
+	uint32_t aligned_addr = cached_addr & HDA_ALIGN_MASK;
+	uint32_t aligned_size = buf_size & HDA_ALIGN_MASK;
+
+	if (aligned_addr != cached_addr) {
+		return -EINVAL;
+	}
+
+	if (aligned_addr < L2_SRAM_BASE || aligned_addr >= L2_SRAM_BASE + L2_SRAM_SIZE) {
+		return -EINVAL;
+	}
+
+	if (aligned_size != buf_size) {
+		return -EINVAL;
+	}
+
+	if (aligned_addr + aligned_size >= L2_SRAM_BASE + L2_SRAM_SIZE) {
+		return -EINVAL;
+	}
+
+	if (*DGCS(base, sid) & DGCS_GEN) {
+		return -EBUSY;
+	}
+
+	if (*DGCS(base, sid) & DGCS_GBUSY) {
+		return -EBUSY;
+	}
+
+	*DGBBA(base, sid) = aligned_addr;
+	*DGBS(base, sid) = aligned_size;
+
+	return 0;
+}
+
+/**
+ * @brief Enable the stream
+ *
+ * @param hda HDA stream set
+ * @param sid Stream ID
+ */
+static inline void cavs_hda_enable(uint32_t base, uint32_t sid)
+{
+	*DGCS(base, sid) |= DGCS_GEN | DGCS_FIFORDY;
+}
+
+/**
+ * @brief Disable stream
+ *
+ * @param hda HDA stream set
+ * @param sid Stream ID
+ */
+static inline void cavs_hda_disable(uint32_t base, uint32_t sid)
+{
+	*DGCS(base, sid) &= ~(DGCS_GEN | DGCS_FIFORDY);
+}
+
+/**
+ * @brief Determine the number of unused bytes in the fifo
+ *
```

Comment:
Pedantry, but you should clarify that this is only useful for an **input**[1] stream ID.  It returns the wrong sense of subtraction if you're trying to figure out how long until an stream inbound to the DSP fills up.


[1] HDA terminology.  "Input" from the host side, output from the DSP, of course.

Reply:
Definitely, in the host output side it'd be useful to see the number of used bytes instead. SOF combines the two into I think something like hda_usable_bytes() which I could also do here.